File: typescript/base/src/main/vsdk/toolkit/render/raytracing/ParallelRaytracer.ts
import { IllegalArgumentException } from "../../../../java/lang/IllegalArgumentException.js";
import { IllegalStateException } from "../../../../java/lang/IllegalStateException.js";
import { platformPrintln } from "../../../../java/lang/_PlatformConsole.js";
import type { WorkerExecutor, WorkerTransferValue } from "../../../../java/concurrent/WorkerProtocol.js";
import type { ArrayList } from "../../../../java/util/ArrayList.js";
import { ConcurrentLinkedQueue } from "../../../../java/util/concurrent/ConcurrentLinkedQueue.js";
import type { RendererConfiguration } from "../../environment/material/RendererConfiguration.js";
import { ParallelProgressMonitorConsumer } from "../../gui/feedback/parallel/ParallelProgressMonitorConsumer.js";
import { ParallelProgressMonitorEvent } from "../../gui/feedback/parallel/ParallelProgressMonitorEvent.js";
import { ParallelProgressMonitorProducer } from "../../gui/feedback/parallel/ParallelProgressMonitorProducer.js";
import type { RGBImageUncompressed } from "../../media/RGBImageUncompressed.js";
import { ZBuffer } from "../../media/ZBuffer.js";
import { DepthBufferMode } from "./DepthBufferMode.js";
import type { RasterTileArea } from "./RasterTileArea.js";
import { RasterTileGenerationStrategy } from "./RasterTileGenerationStrategy.js";
import { RasterTileGenerator } from "./RasterTileGenerator.js";

/**
One band of the image, as the owner asks a worker to render it (see
`ParallelRaytracerTileRenderer`).
*/
export interface ParallelRaytracerTileRequest {
    /** Whatever the worker needs to rebuild the scene (i.e. a scene file name) */
    readonly sceneDescriptor: WorkerTransferValue;
    /** Changes on every `execute`: a worker rebuilds the scene when it changes */
    readonly sceneVersion: number;
    /** False to render without sending progress notices */
    readonly reportProgress: boolean;
    readonly xSize: number;
    readonly ySize: number;
    /** The only `RendererConfiguration` bits `SimpleRaytracer` reads */
    readonly shadingType: number;
    readonly texture: boolean;
    readonly bumpMap: boolean;
    readonly startX: number;
    readonly startY: number;
    readonly width: number;
    readonly height: number;
    /** Kind of depth to return with the band (a `DepthBufferMode`) */
    readonly depthMode: string;
    /** `depthRange` of `DepthBufferMode.OPENGL_DEPTH` */
    readonly openGlDepthRangeNear: number;
    readonly openGlDepthRangeFar: number;
    readonly [key: string]: WorkerTransferValue;
}

/** The pixels of one rendered band, returned by a worker. */
export interface ParallelRaytracerTileResult {
    readonly startY: number;
    readonly height: number;
    readonly bytes: Uint8Array;
    /** Depth of the rows of the band, top row first, or null without depth mode */
    readonly depth: Float32Array | null;
    readonly [key: string]: WorkerTransferValue;
}

/** Creates one worker; the platform decides how (Web Worker, Node worker...). */
export type ParallelRaytracerWorkerFactory = () => WorkerExecutor<
    ParallelRaytracerTileRequest,
    ParallelRaytracerTileResult
>;

/**
TypeScript counterpart of Java's `vsdk.toolkit.render.raytracing.ParallelRaytracer`.

As in Java, it raytraces an image with as many threads as processors are
available: the image is split in horizontal bands (see `RasterTileGenerator`),
several per thread, and each thread takes the next pending band when it
finishes one, so threads stay busy even if some areas of the image are much
more expensive than others. The threads are created once and reused by every
call, so it suits interactive use; `dispose` stops them.

Differences forced by the platform, where threads are workers without a
shared heap and without a blocking join:
  - The workers come from a `ParallelRaytracerWorkerFactory`, since
    `@vitral/base` does not know which platform it runs on. Each worker module
    installs a `ParallelRaytracerTileRenderer`.
  - The scene snapshot can not be shared, so `execute` receives a
    `sceneDescriptor` from which each worker rebuilds the scene once per
    `execute` (and reuses it for all its bands) instead of the snapshot
    itself.
  - A worker can not poll the owner's queue nor write into the owner's image:
    the owner polls for it, and splices the bytes of each band the worker
    returns. Bands never overlap, so the image is the one Java builds in place.
  - `execute` answers a promise where Java blocks.

Optionally (see `setDepthBufferMode`), each call also exports the depth of
the primary ray of every pixel into a `ZBuffer` of the size of the image:
native ray distances, or depth values normalized as OpenGL / WebGL would
store them for the same camera (with its near and far planes and the
configured depth range), so the raytraced image can be composited with
rasterized geometry (see `DepthBufferMode`). Rows of the depth buffer follow
the rows of the image: row 0 is the top one. Each worker returns the depth of
its band along with its pixels.
*/
export class ParallelRaytracer {
    /** Bands per thread: more bands balance the load better */
    private static readonly BANDS_PER_THREAD = 8;

    private readonly numberOfThreads: number;
    private workers: WorkerExecutor<ParallelRaytracerTileRequest, ParallelRaytracerTileResult>[] | null = null;
    private sceneVersion = 0;
    private depthBufferMode: DepthBufferMode = DepthBufferMode.NONE;
    private openGlDepthRangeNear = 0.0;
    private openGlDepthRangeFar = 1.0;
    /** Depth buffer of the last `execute` call, reused while its size fits */
    private depthBuffer: ZBuffer | null = null;

    /**
    @param workerFactory creates each of the workers
    @param numberOfThreads number of workers to render with (at least 1); by
    default, one per available processor
    */
    public constructor(
        private readonly workerFactory: ParallelRaytracerWorkerFactory,
        numberOfThreads: number = ParallelRaytracer.availableProcessors(),
    ) {
        if (!Number.isInteger(numberOfThreads) || numberOfThreads <= 0) {
            throw new IllegalArgumentException("numberOfThreads must be > 0");
        }
        this.numberOfThreads = numberOfThreads;
    }

    /**
    @return the number of processors the platform reports (`hardwareConcurrency`,
    available both in browsers and in Node), or 1 if it reports none
    */
    public static availableProcessors(): number {
        const count: number | undefined =
            typeof navigator !== "undefined" ? navigator.hardwareConcurrency : undefined;
        return count !== undefined && count > 0 ? count : 1;
    }

    /** @return number of workers used to render */
    public getNumberOfThreads(): number {
        return this.numberOfThreads;
    }

    /**
    Selects whether (and how) the next `execute` calls export a depth buffer.
    @param mode kind of depth buffer; null is taken as `NONE`
    */
    public setDepthBufferMode(mode: DepthBufferMode | null): void {
        this.depthBufferMode = mode === null ? DepthBufferMode.NONE : mode;
        if (this.depthBufferMode === DepthBufferMode.NONE) {
            this.depthBuffer = null;
        }
    }

    /** @return kind of depth buffer exported by `execute` */
    public getDepthBufferMode(): DepthBufferMode {
        return this.depthBufferMode;
    }

    /**
    Sets the mapping from normalized device depth to window depth used by
    `DepthBufferMode.OPENGL_DEPTH`, as `depthRange`. By default [0, 1].
    @param near window depth of the near plane
    @param far window depth of the far plane
    */
    public setOpenGlDepthRange(near: number, far: number): void {
        this.openGlDepthRangeNear = near;
        this.openGlDepthRangeFar = far;
    }

    /** @return window depth of the near plane for `DepthBufferMode.OPENGL_DEPTH` */
    public getOpenGlDepthRangeNear(): number {
        return this.openGlDepthRangeNear;
    }

    /** @return window depth of the far plane for `DepthBufferMode.OPENGL_DEPTH` */
    public getOpenGlDepthRangeFar(): number {
        return this.openGlDepthRangeFar;
    }

    /**
    @return depth buffer filled by the last `execute` call, of the size of
    its image, or null if the depth buffer mode is `NONE` or nothing was
    rendered yet. It is reused (overwritten) by the next calls.
    */
    public getDepthBuffer(): ZBuffer | null {
        return this.depthBuffer;
    }

    private getWorkers(): WorkerExecutor<ParallelRaytracerTileRequest, ParallelRaytracerTileResult>[] {
        if (this.workers === null) {
            this.workers = [];
            for (let i = 0; i < this.numberOfThreads; i++) {
                this.workers.push(this.workerFactory());
            }
        }
        return this.workers;
    }

    /**
    Raytraces the scene into the whole image, resolving when it is complete.
    @param resultingImage image to fill; its size gives the resolution
    @param rendererConfiguration quality settings
    @param sceneDescriptor what each worker needs to rebuild the scene
    @param reportProgress true to report in the console the number of
    threads and the progress
    */
    public async execute(
        resultingImage: RGBImageUncompressed,
        rendererConfiguration: RendererConfiguration,
        sceneDescriptor: WorkerTransferValue,
        reportProgress: boolean,
    ): Promise<void>;
    /**
    Raytraces the scene into the whole image and into a depth buffer given by
    the caller, independently of the configured depth buffer mode.
    @param resultingImage image to fill; its size gives the resolution
    @param outDepth depth buffer of the size of the image to fill
    @param depthMode kind of depth values to write; `NONE` leaves `outDepth`
    untouched
    @param rendererConfiguration quality settings
    @param sceneDescriptor what each worker needs to rebuild the scene
    @param reportProgress true to report in the console the number of
    threads and the progress
    */
    public async execute(
        resultingImage: RGBImageUncompressed,
        outDepth: ZBuffer | null,
        depthMode: DepthBufferMode,
        rendererConfiguration: RendererConfiguration,
        sceneDescriptor: WorkerTransferValue,
        reportProgress: boolean,
    ): Promise<void>;
    public async execute(
        resultingImage: RGBImageUncompressed,
        ...args: (ZBuffer | null | DepthBufferMode | RendererConfiguration | WorkerTransferValue | boolean)[]
    ): Promise<void> {
        let outDepth: ZBuffer | null;
        let depthMode: DepthBufferMode;
        let rendererConfiguration: RendererConfiguration;
        let sceneDescriptor: WorkerTransferValue;
        let reportProgress: boolean;

        if (args.length === 5) {
            outDepth = args[0] as ZBuffer | null;
            depthMode = args[1] as DepthBufferMode;
            rendererConfiguration = args[2] as RendererConfiguration;
            sceneDescriptor = args[3] as WorkerTransferValue;
            reportProgress = args[4] as boolean;
            if (outDepth === null) {
                depthMode = DepthBufferMode.NONE;
            }
            if (depthMode !== DepthBufferMode.NONE && outDepth !== null &&
                (outDepth.getXSize() !== resultingImage.getXSize() ||
                 outDepth.getYSize() !== resultingImage.getYSize())) {
                throw new IllegalArgumentException("Depth buffer size must match the image size");
            }
        } else {
            rendererConfiguration = args[0] as RendererConfiguration;
            sceneDescriptor = args[1] as WorkerTransferValue;
            reportProgress = args[2] as boolean;
            depthMode = this.depthBufferMode;
            outDepth = null;
            if (depthMode !== DepthBufferMode.NONE && resultingImage.getXSize() > 0 &&
                resultingImage.getYSize() > 0) {
                if (this.depthBuffer === null ||
                    this.depthBuffer.getXSize() !== resultingImage.getXSize() ||
                    this.depthBuffer.getYSize() !== resultingImage.getYSize()) {
                    this.depthBuffer = new ZBuffer(resultingImage.getXSize(), resultingImage.getYSize());
                }
                outDepth = this.depthBuffer;
            }
        }
        await this.executeWithDepth(
            resultingImage,
            depthMode === DepthBufferMode.NONE ? null : outDepth,
            depthMode,
            rendererConfiguration,
            sceneDescriptor,
            reportProgress,
        );
    }

    private async executeWithDepth(
        resultingImage: RGBImageUncompressed,
        outDepth: ZBuffer | null,
        depthMode: DepthBufferMode,
        rendererConfiguration: RendererConfiguration,
        sceneDescriptor: WorkerTransferValue,
        reportProgress: boolean,
    ): Promise<void> {
        if (resultingImage.getXSize() <= 0 || resultingImage.getYSize() <= 0) {
            return;
        }
        const tileGenerator: RasterTileGenerator = new RasterTileGenerator(
            RasterTileGenerationStrategy.LINEAR,
            resultingImage,
            resultingImage.getXSize(),
            resultingImage.getYSize(),
            this.numberOfThreads * ParallelRaytracer.BANDS_PER_THREAD,
        );
        const generatedTiles: ArrayList<RasterTileArea> = tileGenerator.getTiles();
        const pendingTiles: ConcurrentLinkedQueue<RasterTileArea> = new ConcurrentLinkedQueue<RasterTileArea>(
            generatedTiles,
        );

        let producer: ParallelProgressMonitorProducer | null = null;
        let consumerThread: Promise<void> | null = null;
        // The scene may have changed since the previous image
        this.sceneVersion++;

        if (reportProgress) {
            platformPrintln("Starting parallel raytracing with " + this.numberOfThreads + " threads.");
            const progressEvents: ConcurrentLinkedQueue<ParallelProgressMonitorEvent> =
                new ConcurrentLinkedQueue<ParallelProgressMonitorEvent>();
            producer = new ParallelProgressMonitorProducer(progressEvents);
            producer.init(ParallelRaytracer.calculateTotalProgressElements(generatedTiles));
            consumerThread = new ParallelProgressMonitorConsumer(progressEvents).run();
        }

        try {
            await Promise.all(
                this.getWorkers().map((worker) =>
                    this.drainTiles(
                        worker,
                        pendingTiles,
                        resultingImage,
                        outDepth,
                        outDepth === null ? DepthBufferMode.NONE : depthMode,
                        rendererConfiguration,
                        sceneDescriptor,
                        producer,
                    ),
                ),
            );
            if (!pendingTiles.isEmpty()) {
                throw new IllegalStateException("Parallel raytracing finished with pending tiles");
            }
        } finally {
            if (producer !== null && consumerThread !== null) {
                producer.finish();
                await consumerThread;
            }
        }
    }

    /** Stops the workers. The next `execute` creates them again. */
    public dispose(): void {
        if (this.workers !== null) {
            for (const worker of this.workers) {
                worker.terminate();
            }
            this.workers = null;
        }
    }

    private static calculateTotalProgressElements(generatedTiles: ArrayList<RasterTileArea>): bigint {
        let totalElements = 0n;

        for (let i = 0; i < generatedTiles.size(); i++) {
            totalElements += BigInt(generatedTiles.get(i).getHeight());
        }
        return totalElements;
    }

    /**
    Counterpart of Java's `TileWorker.call()`: keep taking bands off the shared
    queue until it runs dry. The polling happens here rather than inside the
    worker because the queue lives in this heap.
    */
    private async drainTiles(
        worker: WorkerExecutor<ParallelRaytracerTileRequest, ParallelRaytracerTileResult>,
        pendingTiles: ConcurrentLinkedQueue<RasterTileArea>,
        resultingImage: RGBImageUncompressed,
        outDepth: ZBuffer | null,
        depthMode: DepthBufferMode,
        rendererConfiguration: RendererConfiguration,
        sceneDescriptor: WorkerTransferValue,
        progressReporter: ParallelProgressMonitorProducer | null,
    ): Promise<void> {
        let tile: RasterTileArea | undefined;

        while ((tile = pendingTiles.poll()) !== undefined) {
            if (tile.getStartX() !== 0 || tile.getWidth() !== resultingImage.getXSize()) {
                throw new IllegalStateException(
                    "The worker band transfer only supports full-width tiles, which is what " +
                        "RasterTileGenerationStrategy.LINEAR produces",
                );
            }
            const result: ParallelRaytracerTileResult = await worker.execute(
                {
                    sceneDescriptor,
                    sceneVersion: this.sceneVersion,
                    reportProgress: progressReporter !== null,
                    xSize: resultingImage.getXSize(),
                    ySize: resultingImage.getYSize(),
                    shadingType: rendererConfiguration.getShadingType(),
                    texture: rendererConfiguration.isTextureSet(),
                    bumpMap: rendererConfiguration.isBumpMapSet(),
                    startX: tile.getStartX(),
                    startY: tile.getStartY(),
                    width: tile.getWidth(),
                    height: tile.getHeight(),
                    depthMode: depthMode,
                    openGlDepthRangeNear: this.openGlDepthRangeNear,
                    openGlDepthRangeFar: this.openGlDepthRangeFar,
                },
                progressReporter === null ? undefined : { onNotice: () => progressReporter.update(0, 0, 0) },
            );
            ParallelRaytracer.spliceBand(resultingImage, result);
            if (outDepth !== null && result.depth !== null) {
                outDepth.getZBuffer().set(result.depth, result.startY * outDepth.getXSize());
            }
        }
    }

    /** Copies a rendered band back into the shared result image. */
    private static spliceBand(resultingImage: RGBImageUncompressed, result: ParallelRaytracerTileResult): void {
        const ySize: number = resultingImage.getYSize();
        const rowStride: number = resultingImage.getXSize() * 3;
        const raw: Uint8Array = resultingImage.getRawImageDirectBuffer();
        const start: number = (ySize - (result.startY + result.height)) * rowStride;
        raw.set(result.bytes, start);
    }
}

File: java/base/src/main/vsdk/toolkit/render/raytracing/ParallelRaytracer.java
package vsdk.toolkit.render.raytracing;

// Java basic classes
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

// VSDK classes
import vsdk.toolkit.environment.material.RendererConfiguration;
import vsdk.toolkit.environment.scene.SimpleSceneSnapshot;
import vsdk.toolkit.gui.feedback.ProgressMonitor;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorConsumer;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorEvent;
import vsdk.toolkit.gui.feedback.parallel.ParallelProgressMonitorProducer;
import vsdk.toolkit.media.RGBImageUncompressed;
import vsdk.toolkit.media.ZBuffer;

/**
Raytraces an image with as many threads as processors are available to the
virtual machine. The image is split in horizontal bands (see
`RasterTileGenerator`), several per thread, and each thread takes the next
pending band when it finishes one, so threads stay busy even if some areas of
the image are much more expensive than others. Each thread uses its own
`SimpleRaytracer` over the same (read only) scene snapshot, and writes
disjoint rows of the image.

The threads are created once and reused by every call, so this class suits
interactive use (one image per frame). They are daemon threads; `dispose`
stops them.

Optionally (see `setDepthBufferMode`), each call also exports the depth of
the primary ray of every pixel into a `ZBuffer` of the size of the image:
native ray distances, or depth values normalized as OpenGL would store them
for the same camera (with its near and far planes and the configured
`glDepthRange`), so the raytraced image can be composited with rasterized
geometry (see `DepthBufferMode`). Rows of the depth buffer follow the rows of
the image: row 0 is the top one.
*/
public class ParallelRaytracer
{
    /** Bands per thread: more bands balance the load better */
    private static final int BANDS_PER_THREAD = 8;

    private final int numberOfThreads;
    private ExecutorService executorService;

    private DepthBufferMode depthBufferMode;
    private double openGlDepthRangeNear;
    private double openGlDepthRangeFar;
    /// Depth buffer of the last `execute` call, reused while its size fits
    private ZBuffer depthBuffer;

    /**
    Creates a raytracer with one thread per available processor.
    */
    public ParallelRaytracer()
    {
        this(Runtime.getRuntime().availableProcessors());
    }

    /**
    @param numberOfThreads number of threads to render with (at least 1)
    */
    public ParallelRaytracer(int numberOfThreads)
    {
        if ( numberOfThreads <= 0 ) {
            throw new IllegalArgumentException("numberOfThreads must be > 0");
        }
        this.numberOfThreads = numberOfThreads;
        this.executorService = null;
        this.depthBufferMode = DepthBufferMode.NONE;
        this.openGlDepthRangeNear = 0.0;
        this.openGlDepthRangeFar = 1.0;
        this.depthBuffer = null;
    }

    /**
    Selects whether (and how) the next `execute` calls export a depth buffer.
    @param mode kind of depth buffer; null is taken as `NONE`
    */
    public synchronized void setDepthBufferMode(DepthBufferMode mode)
    {
        depthBufferMode = (mode == null) ? DepthBufferMode.NONE : mode;
        if ( depthBufferMode == DepthBufferMode.NONE ) {
            depthBuffer = null;
        }
    }

    /**
    @return kind of depth buffer exported by `execute`
    */
    public synchronized DepthBufferMode getDepthBufferMode()
    {
        return depthBufferMode;
    }

    /**
    Sets the mapping from normalized device depth to window depth used by
    `DepthBufferMode.OPENGL_DEPTH`, as `glDepthRange`. By default [0, 1].
    @param near window depth of the near plane
    @param far window depth of the far plane
    */
    public synchronized void setOpenGlDepthRange(double near, double far)
    {
        openGlDepthRangeNear = near;
        openGlDepthRangeFar = far;
    }

    /**
    @return window depth of the near plane for `DepthBufferMode.OPENGL_DEPTH`
    */
    public synchronized double getOpenGlDepthRangeNear()
    {
        return openGlDepthRangeNear;
    }

    /**
    @return window depth of the far plane for `DepthBufferMode.OPENGL_DEPTH`
    */
    public synchronized double getOpenGlDepthRangeFar()
    {
        return openGlDepthRangeFar;
    }

    /**
    @return depth buffer filled by the last `execute` call, of the size of
    its image, or null if the depth buffer mode is `NONE` or nothing was
    rendered yet. It is reused (overwritten) by the next calls.
    */
    public synchronized ZBuffer getDepthBuffer()
    {
        return depthBuffer;
    }

    /**
    @return number of threads used to render
    */
    public int getNumberOfThreads()
    {
        return numberOfThreads;
    }

    private synchronized ExecutorService getExecutorService()
    {
        if ( executorService == null ) {
            AtomicInteger threadCount = new AtomicInteger();
            ThreadFactory factory = runnable -> {
                Thread thread = new Thread(runnable,
                    "parallel-raytracer-" + threadCount.incrementAndGet());
                thread.setDaemon(true);
                return thread;
            };
            executorService = Executors.newFixedThreadPool(numberOfThreads, factory);
        }
        return executorService;
    }

    /**
    Raytraces the scene into the whole image, returning when it is complete.
    @param resultingImage image to fill; its size gives the resolution
    @param rendererConfiguration quality settings
    @param sceneSnapshot scene to render, seen from its camera snapshot
    @param reportProgress true to report in the console the number of
    threads and the progress
    */
    public void execute(RGBImageUncompressed resultingImage,
                        RendererConfiguration rendererConfiguration,
                        SimpleSceneSnapshot sceneSnapshot,
                        boolean reportProgress)
    {
        DepthBufferEncoder depthEncoder = null;
        ZBuffer outDepth = null;

        if ( resultingImage.getXSize() <= 0 || resultingImage.getYSize() <= 0 ) {
            return;
        }
        synchronized ( this ) {
            if ( depthBufferMode != DepthBufferMode.NONE ) {
                if ( depthBuffer == null ||
                     depthBuffer.getXSize() != resultingImage.getXSize() ||
                     depthBuffer.getYSize() != resultingImage.getYSize() ) {
                    depthBuffer = new ZBuffer(resultingImage.getXSize(),
                        resultingImage.getYSize());
                }
                outDepth = depthBuffer;
                depthEncoder = new DepthBufferEncoder(depthBufferMode,
                    sceneSnapshot.getCameraSnapshot(), openGlDepthRangeNear, openGlDepthRangeFar);
            }
        }
        execute(resultingImage, outDepth, depthEncoder, rendererConfiguration,
            sceneSnapshot, reportProgress);
    }

    /**
    Raytraces the scene into the whole image and into a depth buffer given by
    the caller, independently of the configured depth buffer mode.
    @param resultingImage image to fill; its size gives the resolution
    @param outDepth depth buffer of the size of the image to fill
    @param depthMode kind of depth values to write; `NONE` leaves `outDepth`
    untouched
    @param rendererConfiguration quality settings
    @param sceneSnapshot scene to render, seen from its camera snapshot
    @param reportProgress true to report in the console the number of
    threads and the progress
    */
    public void execute(RGBImageUncompressed resultingImage,
                        ZBuffer outDepth,
                        DepthBufferMode depthMode,
                        RendererConfiguration rendererConfiguration,
                        SimpleSceneSnapshot sceneSnapshot,
                        boolean reportProgress)
    {
        DepthBufferEncoder depthEncoder = null;

        if ( outDepth != null && depthMode != null && depthMode != DepthBufferMode.NONE ) {
            if ( outDepth.getXSize() != resultingImage.getXSize() ||
                 outDepth.getYSize() != resultingImage.getYSize() ) {
                throw new IllegalArgumentException(
                    "Depth buffer size must match the image size");
            }
            depthEncoder = new DepthBufferEncoder(depthMode,
                sceneSnapshot.getCameraSnapshot(),
                getOpenGlDepthRangeNear(), getOpenGlDepthRangeFar());
        }
        execute(resultingImage, outDepth, depthEncoder, rendererConfiguration,
            sceneSnapshot, reportProgress);
    }

    private void execute(RGBImageUncompressed resultingImage,
                         ZBuffer outDepth,
                         DepthBufferEncoder depthEncoder,
                         RendererConfiguration rendererConfiguration,
                         SimpleSceneSnapshot sceneSnapshot,
                         boolean reportProgress)
    {
        if ( resultingImage.getXSize() <= 0 || resultingImage.getYSize() <= 0 ) {
            return;
        }
        RasterTileGenerator tileGenerator = new RasterTileGenerator(
            RasterTileGenerationStrategy.LINEAR,
            resultingImage,
            resultingImage.getXSize(),
            resultingImage.getYSize(),
            numberOfThreads * BANDS_PER_THREAD);
        List<RasterTileArea> generatedTiles = tileGenerator.getTiles();
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles =
            new ConcurrentLinkedQueue<>(generatedTiles);

        ParallelProgressMonitorProducer producer = null;
        Thread consumerThread = null;

        if ( reportProgress ) {
            System.out.println(
                "Starting parallel raytracing with " + numberOfThreads + " threads.");
            ConcurrentLinkedQueue<ParallelProgressMonitorEvent> progressEvents =
                new ConcurrentLinkedQueue<>();
            producer = new ParallelProgressMonitorProducer(progressEvents);
            consumerThread = new Thread(
                new ParallelProgressMonitorConsumer(progressEvents),
                "parallel-progress-monitor-consumer");
            producer.init(calculateTotalProgressElements(generatedTiles));
            consumerThread.start();
        }

        try {
            List<Future<Void>> futures = new ArrayList<>(numberOfThreads);
            for ( int i = 0; i < numberOfThreads; i++ ) {
                futures.add(getExecutorService().submit(new TileWorker(
                    pendingTiles,
                    resultingImage,
                    depthEncoder != null ? outDepth : null,
                    depthEncoder,
                    rendererConfiguration,
                    sceneSnapshot,
                    producer)));
            }
            for ( Future<Void> future : futures ) {
                future.get();
            }
            if ( !pendingTiles.isEmpty() ) {
                throw new IllegalStateException("Parallel raytracing finished with pending tiles");
            }
        }
        catch ( InterruptedException e ) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Parallel raytracing was interrupted", e);
        }
        catch ( ExecutionException e ) {
            throw new IllegalStateException("Parallel raytracing failed", e);
        }
        finally {
            if ( producer != null ) {
                producer.finish();
                try {
                    consumerThread.join();
                }
                catch ( InterruptedException e ) {
                    Thread.currentThread().interrupt();
                }
            }
        }
    }

    /**
    Stops the threads. The next `execute` creates them again.
    */
    public synchronized void dispose()
    {
        if ( executorService != null ) {
            executorService.shutdownNow();
            executorService = null;
        }
    }

    private static long calculateTotalProgressElements(List<RasterTileArea> generatedTiles)
    {
        long totalElements = 0;

        for ( RasterTileArea tile : generatedTiles ) {
            totalElements += tile.getHeight();
        }
        return totalElements;
    }

    private record TileWorker(
        ConcurrentLinkedQueue<RasterTileArea> pendingTiles,
        RGBImageUncompressed resultingImage,
        ZBuffer depthBuffer,
        DepthBufferEncoder depthEncoder,
        RendererConfiguration rendererConfiguration,
        SimpleSceneSnapshot sceneSnapshot,
        ProgressMonitor progressReporter)
        implements Callable<Void> {

        @Override
        public Void call()
        {
            RasterTileArea tile;
            SimpleRaytracer raytracer = new SimpleRaytracer();

            while ( (tile = pendingTiles.poll()) != null ) {
                raytracer.execute(
                    resultingImage,
                    rendererConfiguration,
                    sceneSnapshot,
                    progressReporter,
                    depthBuffer,
                    depthEncoder,
                    tile.getStartX(),
                    tile.getStartY(),
                    tile.getEndX(),
                    tile.getEndY());
            }

            return null;
        }
    }
}
